bestofjs/bestofjs
# Refactor pagination / search feature

## Goal

- Improve search / pagination feature, making the code more "DRY".
- Use Zod schema to validate URL query string params  instead of doing validation by ourselves
 
With server components, only top level components (`page.tsx`) can access the search params via the props so the parsing logic is made at the top level, using a specific parsers for project and tag searches.

How it works:

Pagination is considered as a subset of the `search state`.

Global pagination logic is defined in `@/lib/page-search-state` and provides the base state made of `page` (the current page number) and `limit` (number of items per page) properties.

Every entity (projects and tags for now) extend the pagination schema with search features: `sort`, `tags`, `searchQuery`...

- `@/app/projects/project-search-state`: provides the schema for the Project search
- `@/app/projects/tag-search-state`: provides the schema for the Tag search

A `SearchState` parser object is responsible for parsing the search parameters and provide a `parse` method that returns:

- the `searchState`: a plain object that describes the state of the search (including the pagination)
- a `buildPageURL` used to build `<Link>` URLs given changes to be applied to the searchState

The buildPageURL function parameter is a bit similar to the functional variant of React `setState` as it takes the current state and return the next one.

In a nutshell, in `projects/page.tsx`:

```ts
const searchStateParser = new ProjectSearchStateParser();

export default async function ProjectsPage({ searchParams }: PageProps) {
  const { searchState, buildPageURL } = searchStateParser.parse(searchParams);
  const { projects, total } = await fetchPageData(searchState);

  return (
      <ProjectPaginatedList
        projects={projects}
        total={total}
        searchState={searchState}
        buildPageURL={buildPageURL}
    />
  }
}
```

Lower in the component tree, when we need to implement a "Next" button:

```tsx
const nextPageURL = buildPageURL((state) => ({
  ...state,
  page: state.page + 1,
}));

return <Link href={nextPageURL}>Next</Link>
```

Related projects: I had a look at this project https://github.com/47ng/nuqs#accessing-searchparams-in-server-components

## How to test

Ensure search and pagination is not broken:

- From Projects page: `/projects`: click on the dropdown button to change the "sort fillter", navigate between pages...
- From Tags page: `/tags`: click on the drop down button to switch between "By alphabetical order" and "By number of project"

## Screenshots

#### `/projects` page

![image](https://github.com/user-attachments/assets/c34e0b55-6a0d-40f8-a1b5-2e868d40692a)

#### `/tags` page

![image](https://github.com/user-attachments/assets/87975e58-aec1-46f5-bb1b-c3021951beeb)
